Keep sugar lumps above 100

uecasm opened the pull request:
I haven't run the numbers or anything (or tested this change!), but I noticed that when it hits 99 sugar lumps the bot seems to immediately consume 10-20 of them, and I think this is why.

Since there's a CPS bonus for up to 100 lumps, this change makes it pick a slightly different strategy; it will wait until it has 105 lumps before buying something that requires 5 lumps, leaving the total always at 100 or higher.

The first purchases (to unlock minigames and level them to 10) are exempt from this, as before; it will buy those as soon as it can.

uecasm committed:
Keep sugar lumps above 100

prinzstani commented:
Very good idea. 
1) Why would you take +99 and not +100 if you are interested in keeping it at 100?
2) There is in fact a slight challenge here, as sugar lumps are used in other contexts as well: for sugar rush, and for spells. They are only allowed to use lumps if there are at least 101 sugar lumps.
If we now wait until we have 106 for advancing to level 6, then there is a risk that the lumps are first used for spells, which is a bad idea. So we need some more code avoiding the use of lumps for spells as long as there are not all buildings filled to level 10.

uecasm commented:
It's +99 because it's a `<` comparison.  If it needs 1 lump to level then `1+99 < lumps` will fail until you have 101 lumps, so you'll be left with 100 afterwards.

uecasm commented:
Although in saying that, it's possible that I have the values wrong.  If `me.level` is 0 when you're about to spend one lump to level it up, then it should be `+100` instead.

uecasm committed:
Update sugar lump behaviour

uecasm commented:
I've switched it to 100 and added a flag.  Again, I haven't actually tested this.  (Not actually sure how to test it, actually.  I tried loading it from the raw.githubcontent address but it wouldn't load into the game for some reason.)

Not sure if I missed something, but I only found a couple of places where it seemed to use sugar lumps outside of buying building levels.  One is for an achievement, so I left it alone.  The other is after casting "conjure baked goods", and this change should disable doing that while it's still trying to level buildings.

Edit: oh right, there's also Sugar Frenzy.  But that's only bought once per run, after 3 days, and only takes one lump.  I think that's probably safe to leave enabled.

prinzstani commented:
That looks promising - will try some tests.
The most easy way to test is to load it into the console after having prepared everything for the test.
I assume there might be some problems when the recursive call is the last call, which would mean it gives the result false almost all the time.
Maybe it is enough to use 
Game.ObjectsById.every(function (b) { return b.level>9; })
to check that all buildings are at level 10, before conjuring the baked goods.

prinzstani commented:
And I agree, we do not need to care for the sugar frenzy, as it is not activated often.

uecasm commented:
I tried `Game.LoadMod('https://raw.githubusercontent.com/uecasm/CookieBot/patch-2/cookieAutoPlayBeta.js');` at the console, but it gets CORB blocked.  Or do you mean pasting the whole content of the script directly?

There shouldn't be any issues with the last call; that's why it sets the value to true outside of the loop and only sets it to false if some building isn't level 10 yet.  And yes, doing the level check directly in Conjure Baked Goods would also work, but this seemed more future-proof.

prinzstani commented:
Yes, I normallypaste the whole thing in when testing level 0. After that, I put the code into the beta and do some more testing before putting into the release.
I have now merged your changes - so feel free to do some more testing.

prinzstani commented:
Testing OK, so we keep the code like this.
